File: algorithm/integer-replacement/integer-replacement.cpp
class Solution {
    int min(int a, int b) {
        return a <= b ? a : b;
    }
public:
    int integerReplacement(int n) {
        
        unsigned int x = n;
        
        if (n == 2)
            return 1;
        else if (n == 1)
            return 0;

        if (x % 2)
            return 1 + min(integerReplacement(x + 1), integerReplacement(x - 1));
        else
            return 1 + integerReplacement(x / 2);
    }
};

File: algorithm/integer-replacement/integer-replacement.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
__date__= ' 9 12, 2016 '
__author__= 'mkfsn'


class Solution(object):
    def integerReplacement(self, n):
        """
        :type n: int
        :rtype: int
        """

        if n == 2:
            return 1
        elif n == 1:
            return 0

        if (n % 2):
            plus = self.integerReplacement(n + 1)
            minus = self.integerReplacement(n - 1)
            return 1 + min([plus, minus])
        else:
            return 1 + self.integerReplacement(n / 2)

s = Solution()
print s.integerReplacement(2147483647)
